### PR TITLE
BUG: Remove complex floor divide

### LINKED ISSUE #13236: np.floor_divide drops imaginary part silently
### Reproducing code example:
```
import numpy as np
a = np.arange(10) + 1j* np.arange(10)
a
# array([0.+0.j, 1.+1.j, 2.+2.j, 3.+3.j, 4.+4.j, 5.+5.j, 6.+6.j, 7.+7.j,
#           8.+8.j, 9.+9.j])
a / 2
# array([0. +0.j , 0.5+0.5j, 1. +1.j , 1.5+1.5j, 2. +2.j , 2.5+2.5j,
#          3. +3.j , 3.5+3.5j, 4. +4.j , 4.5+4.5j])
a // 2
# array([0.+0.j, 0.+0.j, 1.+0.j, 1.+0.j, 2.+0.j, 2.+0.j, 3.+0.j, 3.+0.j,
#          4.+0.j, 4.+0.j])
```
Is there a special reason for this behavior? I would expect to floor_divide the imaginary and complex part.

### Error message:

No error or warning is thrown, not even that imaginary part is getting lost.

### Numpy/Python version information:
1.16.1 
3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0]

### PR DESCRIPTION
**Pull request by ganesh-k13**
### Changes:
1. Removed `floor_divide` for all complex types
2. Modified UFuncs `floor_divde` TC to check for `TypeError`. Added same for divmod and remainder.
3. Modified `test_core` to check `TypeError` raised for unsupported types.
4. Removed Typing Tests

### TODO:
- [x] Add release note (Need PR number for this) 

closes: #13236
finishes: #13245 (work done by @kaivu1999)

### CODE REVIEW

File: numpy/ma/tests/test_core.py [hunk 1]
```diff
@@ -1317,7 +1317,7 @@ def test_minmax_dtypes(self):
                               dtype=float_dtype)
             assert_equal(zm.min(), float_dtype(-np.inf-1j))
             assert_equal(zm.max(), float_dtype(np.inf+2j))
-
```
Comment (ganesh-k13) [inline numpy/ma/tests/test_core.py:1320]: I really need to change my .vimrc one day

**Comment by ganesh-k13** [comment]
@kaivu1999 didn't cross my mind to check with you before raising this. If you see this before the review for this code is done, please feel free to re-open your PR add the extra changes I have made to fix the tests.

**Review by BvB93 — COMMENTED** [review]
There are a couple of places in the stub files that still contain references to complex floordivision:
https://github.com/numpy/numpy/blob/9fcc13258960adb94d7dda0b478c0997763d4c4d/numpy/__init__.pyi#L2310-L2311
https://github.com/numpy/numpy/blob/9fcc13258960adb94d7dda0b478c0997763d4c4d/numpy/__init__.pyi#L2338-L2339
https://github.com/numpy/numpy/blob/9fcc13258960adb94d7dda0b478c0997763d4c4d/numpy/__init__.pyi#L3258-L3259

The same holds for the typing tests in [reveal/arithmetic.py](https://github.com/numpy/numpy/blob/main/numpy/typing/tests/data/reveal/arithmetic.py) and [pass/arithmetic.py](https://github.com/numpy/numpy/blob/main/numpy/typing/tests/data/pass/arithmetic.py).

**Comment by ganesh-k13** [comment]
Thanks @BvB93, I'll remove those as well. I wanted to understand how and where the three references in `numpy/numpy/__init__.pyi` are being used. When I tried removing them for unsigned int, I was still able to go forward with the division. So where are they used?

File: numpy/ma/tests/test_core.py [hunk 2]
```diff
@@ -2853,39 +2853,51 @@ def test_inplace_multiplication_array_type(self):
 
     def test_inplace_floor_division_scalar_type(self):
         # Test of inplace division
+        # Check for TypeError in case of unsupported types
+        unsupported = {np.complex64, np.complex128, np.complex256}
```
Comment (ganesh-k13) [inline numpy/ma/tests/test_core.py:2857]: Ohh seems like `complex256` is not there on all architectures. Will fix this.

Comment (BvB93) [reply numpy/ma/tests/test_core.py:2857]: You can try iterating over `np.typecodes["Complex"]`, which will yield the platform-agnostic type-codes of all complex types.

**Comment by BvB93** [comment]
> When I tried removing them for unsigned int, I was still able to go forward with the division. So where are they used?

Yeah, that makes sense actually. Each of these `_ArrayLike<x>_co` variables represents an union of _all_ types that can be cast into `<x>`, according to the rules of `casting="same_kind"`. So removing the overload for unsigned integers means that, next time, they'll be handled by the overload for signed integers. Remove that overload as well and they'll be handled by the float overload, _etc._. 

Complex is at the end of this tree though (as it cannot be safely cast into any other type), so removing it means that complex operations are no longer allowed.

https://github.com/numpy/numpy/blob/9fcc13258960adb94d7dda0b478c0997763d4c4d/numpy/__init__.pyi#L2016-L2021

**Comment by ganesh-k13** [comment]
Thanks! Always wondered where the casting order is taken care for arrays.

File: numpy/ma/tests/test_core.py (hunk 2 above)
Comment (ganesh-k13) [reply numpy/ma/tests/test_core.py]: Thanks! It worked

File: numpy/ma/tests/test_core.py [hunk 3]
```diff
@@ -2853,39 +2853,51 @@ def test_inplace_multiplication_array_type(self):
 
     def test_inplace_floor_division_scalar_type(self):
         # Test of inplace division
+        # Check for TypeError in case of unsupported types
+        unsupported = {np.sctypeDict[t] for t in np.typecodes["Complex"]}
```
Comment (BvB93) [inline numpy/ma/tests/test_core.py:2857]:
```suggestion
        unsupported = {np.dtype(t).type for t in np.typecodes["Complex"]}
```
Style suggestion; the same applies to `test_inplace_floor_division_array_type`.

Comment (ganesh-k13) [reply numpy/ma/tests/test_core.py]: Thanks, I have updated in same commit.

**Comment by ganesh-k13** [comment]
Hey folks, any changes needed here?

**Review by mattip — APPROVED** [review]
Do we know if this will have implications for scipy/astropy/pandas?

**Comment by ganesh-k13** [comment]
Will building them with this branch of NumPy help? Will be happy to do so.

**Comment by charris** [comment]
Thanks @ganesh-k13 .Lets get this in for testing at least.

**Pull request merged**